capreolus/trainer/__init__.py

```diff
@@ -292,3 +322,265 @@ def predict(self, reranker, pred_data, pred_fn):
+            # Skipping dumping metrics and plotting loss since that should be done through tensorboard
+
+    def eval_and_save_best_model(self, reranker, dev_records, train_output_path, dev_output_path, dev_best_metric, qrels, metric, niter):
+        """
+        Attempt early stopping
+        """
+        dev_best_weight_fn, weights_output_path, info_output_path, loss_fn = self.get_paths_for_early_stopping(
+            train_output_path, dev_output_path
+        )
+
+        pred_fn = dev_output_path / f"{niter}.run"
+        preds = self.predict(reranker, dev_records, pred_fn)
+
+        # log dev metrics
+        metrics = evaluator.eval_runs(preds, qrels, ["ndcg_cut_20", "map", "P_20"])
+        logger.info("dev metrics: %s", " ".join([f"{metric}={v:0.3f}" for metric, v in sorted(metrics.items())]))
+
+        tf.summary.scalar('ndcg20', metrics['ndcg_cut_20'], step=niter)
+        tf.summary.scalar('map', metrics["map"], step=niter)
+        tf.summary.scalar('P_20', metrics['P_20'], step=niter)
+
+        # write best dev weights to file
+        if metrics[metric] > dev_best_metric:
+            reranker.save_weights(dev_best_weight_fn, self.optimizer)
+
+    def predict(self, reranker, dev_records, pred_fn):
+        pred_dict = defaultdict(lambda: dict())
+        for step, batch in enumerate(dev_records.batch(1)):
+            qid, doc_id = batch['qid'][0].numpy(), batch['posdoc_id'][0].numpy()
+            pred_dict[qid.decode('utf-8')][doc_id.decode('utf-8')] = reranker.test(batch['posdoc'], batch['query'], batch['query_idf']).numpy().astype(np.float16).item()
+
+        os.makedirs(os.path.dirname(pred_fn), exist_ok=True)
+        Searcher.write_trec_run(pred_dict, pred_fn)
+
+        return dict(pred_dict)
+
+    def create_tf_feature(self, qid, query, query_idf, posdoc_id, posdoc, negdoc_id, negdoc):
+        """
+        Creates a single tf.train.Feature instance (i.e, a single sample)
+        """
+        feature = {
+            "qid": tf.train.Feature(bytes_list=tf.train.BytesList(value=[qid.encode('utf-8')])),
+            "query": tf.train.Feature(float_list=tf.train.FloatList(value=query)),
+            "query_idf": tf.train.Feature(float_list=tf.train.FloatList(value=query_idf)),
+            "posdoc_id": tf.train.Feature(bytes_list=tf.train.BytesList(value=[posdoc_id.encode('utf-8')])),
+            "posdoc": tf.train.Feature(float_list=tf.train.FloatList(value=posdoc)),
+        }
+
+        if negdoc_id:
+            feature["negdoc_id"] = tf.train.Feature(bytes_list=tf.train.BytesList(value=[negdoc_id.encode('utf-8')])),
+            feature["negdoc"] = tf.train.Feature(float_list=tf.train.FloatList(value=negdoc))
+
+        return feature
+
+    def write_tf_record_to_file(self, dir_name, tf_features):
+        filename = dir_name / "{}.tfrecord".format(str(uuid.uuid4()))
+        examples = [tf.train.Example(features=tf.train.Features(feature=feature)) for feature in tf_features]
```

Comment:
We might have to ditch generators altogether - if we use generators here, a "PyFunc" gets included in the resulting tf graph and it thus can't be executed on TPUs because a TPU does not have a python interpreter